src/extremeweatherbench/utils.py

```diff
@@ -414,3 +431,175 @@ def read_event_yaml(input_pth: str | Path) -> dict:
+    point_obs_metadata = case_subset_point_obs_df[point_obs_metadata_vars]
+
+    # Reset index to allow for easier modification
+    case_subset_point_obs_df = case_subset_point_obs_df.reset_index()
+    case_subset_point_obs_df.loc[:, "station"] = case_subset_point_obs_df[
+        "station"
+    ].astype("str")
+
+    # Set up multiindex to enable slicing along individual timesteps
+    case_subset_point_obs_df = case_subset_point_obs_df.set_index("time")
+    case_subset_point_obs_df = case_subset_point_obs_df.set_index(
+        "station", append=True
+    )
+    case_subset_point_obs_df = case_subset_point_obs_df.sort_index(
+        level="station"
+    ).swaplevel(0, 1)
+
+    aligned_forecast_list = []
+    aligned_observation_list = []
+
+    # Loop init and lead times to prevent indexing error from duplicate valid times
+    for init_time, lead_time in itertools.product(
+        forecast_da.init_time, forecast_da.lead_time
+    ):
+        valid_time = pd.Timestamp(
+            init_time.values + pd.to_timedelta(lead_time.values, unit="h").to_numpy()
+        )
+        if valid_time in case_subset_point_obs_df.index.levels[1]:
+            obs_timeslice = case_subset_point_obs_df.loc[
+                pd.IndexSlice[:, valid_time], :
+            ].reset_index()
+            station_ids = obs_timeslice["station"]
+            lons = xr.DataArray(obs_timeslice["longitude"].values, dims="station")
+            lats = xr.DataArray(obs_timeslice["latitude"].values, dims="station")
+
+            grid_at_obs_da = forecast_da.sel(
+                init_time=init_time, lead_time=lead_time
+            ).interp(latitude=lats, longitude=lons, method="nearest")
+            grid_at_obs_da = grid_at_obs_da.assign_coords(
+                {"station": station_ids, "time": valid_time}
+            )
+            grid_at_obs_da["lead_time"] = lead_time
+            grid_at_obs_da["init_time"] = init_time
+
+            valid_time_subset_obs_da = obs_timeslice.to_xarray().set_coords(
+                point_obs_metadata
+            )[forecast_da.name]
+            valid_time_subset_obs_da = valid_time_subset_obs_da.swap_dims(
+                {"index": "station"}
+            )
+            valid_time_subset_obs_da["lead_time"] = lead_time
+            valid_time_subset_obs_da["init_time"] = init_time
+
+            aligned_observation_list.append(valid_time_subset_obs_da)
+            aligned_forecast_list.append(grid_at_obs_da)
+
+    # concat the dataarrays along the station dimension
+    interpolated_forecast = xr.concat(aligned_forecast_list, dim="station")
+    interpolated_observation = xr.concat(aligned_observation_list, dim="station")
+    return (interpolated_forecast, interpolated_observation)
+
+
+def derive_indices_from_init_time_and_lead_time(
+    dataset: xr.Dataset,
+    start_date: datetime.datetime,
+    end_date: datetime.datetime,
+) -> Tuple[np.ndarray]:
```

Comment:
Think about this output argtype - this is saying you'll pass back a `tuple` with a single value, an `np.ndarray`. Is that what we want to pass back? Why wouldn't we just directly pass back the `np.ndarray` in that case?

If you mean that this is a variable length output - that it could be a tuple _of one ore more_ `np.ndarrays`, then there is actually a specific way you indicate that:

```
def func(*args) -> Tuple[np.ndarray,...]:
    pass
```

In either case I'd question why we should return a `tuple` instead of a `list`.